File: codingTest/2week-assignment/day3/B11725.py
# https://www.acmicpc.net/problem/11725
# 트리에 부모찾기

def dfs(curr_idx):
    visited[curr_idx] = True

    for i in range(len(graph[curr_idx])):
        next_idx = graph[curr_idx][i]
        if visited[next_idx] == False:
            answer[next_idx] = curr_idx
            dfs(next_idx)


N = int(input())

graph = [None] * (N + 1)
for i in range(1, N + 1):
    graph[i] = []

visited = [False] * (N + 1)
answer = [0] * (N + 1)

for i in range(N - 1):
    x, y = map(int, input().split())
    graph[x].append(y)
    graph[y].append(x)

dfs(1)


for i in range(2,len(answer)):
    print(answer[i])

for i in range(1, N + 1):
    print(f"Node {i}: {graph[i]}")

File: java-version/src/week2/day3/B11725.java
package week2.day3;


//https://www.acmicpc.net/problem/11725

//트리의 부모 찾기


import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.StringTokenizer;

public class B11725 {
    static ArrayList<Integer>[] graph;
    static boolean[] visited;
    static int N;
    static int[] answer;

    static void dfs(int idx) {
        visited[idx] = true;

        for (int i = 0; i < graph[idx].size(); i++) {
            int next = graph[idx].get(i);
            if (!visited[next]) {
                answer[next] = idx;
                dfs(next);
            }
        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(br.readLine());
        graph = new ArrayList[N+1];
        for (int i = 1; i <= N; i++) {
            graph[i] = new ArrayList<>();
        }
        visited = new boolean[N+1 ];
        answer = new int[N+1];

        int x, y;
        for (int i = 0; i < N - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            x = Integer.parseInt(st.nextToken());
            y = Integer.parseInt(st.nextToken());
            graph[x].add(y);
            graph[y].add(x);
        }

        dfs(1);

        for (int i = 2; i <= N; i++) {
            System.out.println(answer[i]);
        }
    }
}